### PR TITLE
evaluator,parser: implement some other builtin functions

### PR DESCRIPTION
**Pull request by zyguan**

### CODE REVIEW

File: evaluator/builtin_string.go [hunk 1]
```diff
@@ -463,6 +463,23 @@ func builtinTrim(args []types.Datum, _ context.Context) (d types.Datum, err erro
 	return d, nil
 }
 
+// For LTRIM & RTRIM
+// See: https://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_ltrim
+// See: https://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_rtrim
+func trimFn(fn func(string, string) string, cutset string) BuiltinFunc {
```
Comment (shenli) [inline evaluator/builtin_string.go]: What the different with function trimLeft and trimRight?

File: parser/parser.y [hunk 2]
```diff
@@ -1837,7 +1840,7 @@ UnReservedKeyword:
 NotKeywordToken:
 	"ABS" | "ADDDATE" | "ADMIN" | "COALESCE" | "CONCAT" | "CONCAT_WS" | "CONNECTION_ID" | "CUR_TIME"| "COUNT" | "DAY"
 |	"DATE_ADD" | "DATE_SUB" | "DAYNAME" | "DAYOFMONTH" | "DAYOFWEEK" | "DAYOFYEAR" | "FOUND_ROWS" | "GROUP_CONCAT"| "HOUR"
-|	"IFNULL" | "LENGTH" | "LOCATE" | "MAX" | "MICROSECOND" | "MIN" | "MINUTE" | "NULLIF" | "MONTH" | "NOW" | "POW"
+|	"IFNULL" | "ISNULL" | "LENGTH" | "LOCATE" | "MAX" | "MICROSECOND" | "MIN" | "MINUTE" | "NULLIF" | "MONTH" | "NOW" | "POW"
```
Comment (shenli) [inline parser/parser.y]: Should we put LTRIM and RTRIM here? So as the "upper".

File: evaluator/builtin_string.go (hunk 1 above)
Comment (zyguan) [inline evaluator/builtin_string.go]: ```trimFn``` here is just a high-order function. It's used to build trimLeft and trimRight. Maybe we could add following lines to make it more clear:
``` go
var (
  builtinLTrim = trimFn(strings.TrimLeft, spaceChars)
  builtinRTrim = trimFn(strings.TrimRight, spaceChars)
)
```

File: parser/parser.y (hunk 2 above)
Comment (zyguan) [inline parser/parser.y]: I'm not very clear about that. As TRIM is in the rule, these tokens might better be there. Sorry for not noticing that.

**Comment by shenli** [comment]
@coocood @zimulala PTAL

**Comment by zxylvlp** [comment]
I think you should do TypeInfer for each function.

**Comment by shenli** [comment]
@zxylvlp +1
@zyguan Please refer: https://github.com/pingcap/tidb/blob/master/optimizer/typeinferer.go#L253

**Comment by zyguan** [comment]
@zxylvlp @shenli Thanks for pointing it out, I will fix it later.

**Comment by zyguan** [comment]
@shenli @zxylvlp PTAL
Did i miss something?

**Comment by shenli** [comment]
Well done!
LGTM

**Comment by zxylvlp** [comment]
LGTM

**Comment by zyguan** [comment]
Thanks, I will rebase this PR on master soon.

**Comment by ngaut** [comment]
PTAL

**Pull request merged**